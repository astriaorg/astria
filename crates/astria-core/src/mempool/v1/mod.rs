pub mod transaction_status;
