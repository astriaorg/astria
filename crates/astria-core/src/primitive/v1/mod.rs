pub mod asset;
pub mod u128;

use base64::{
    display::Base64Display,
    prelude::BASE64_STANDARD,
};
use sha2::{
    Digest as _,
    Sha256,
};

use crate::{
    generated::primitive::v1 as raw,
    Protobuf,
};

pub const ADDRESS_LEN: usize = 20;
pub const ROLLUP_ID_LEN: usize = 32;
pub const FEE_ASSET_ID_LEN: usize = 32;

impl Protobuf for merkle::Proof {
    type Error = merkle::audit::InvalidProof;
    type Raw = raw::Proof;

    fn try_from_raw_ref(raw: &Self::Raw) -> Result<Self, Self::Error> {
        // XXX: Implementing this by cloning is ok because `audit_path`
        //      has to be cloned always due to `UncheckedProof`'s constructor.
        Self::try_from_raw(raw.clone())
    }

    fn try_from_raw(raw: Self::Raw) -> Result<Self, Self::Error> {
        let Self::Raw {
            audit_path,
            leaf_index,
            tree_size,
        } = raw;
        let leaf_index = leaf_index.try_into().expect(
            "running on a machine with at least 64 bit pointer width and can convert from u64 to \
             usize",
        );
        let tree_size = tree_size.try_into().expect(
            "running on a machine with at least 64 bit pointer width and can convert from u64 to \
             usize",
        );
        Self::unchecked()
            .audit_path(audit_path.to_vec())
            .leaf_index(leaf_index)
            .tree_size(tree_size)
            .try_into_proof()
    }

    fn to_raw(&self) -> Self::Raw {
        // XXX: Implementing in terms of clone is ok because the fields would need to be cloned
        // anyway.
        self.clone().into_raw()
    }

    fn into_raw(self) -> Self::Raw {
        let merkle::audit::UncheckedProof {
            audit_path,
            leaf_index,
            tree_size,
        } = self.into_unchecked();
        Self::Raw {
            audit_path: audit_path.into(),
            leaf_index: leaf_index.try_into().expect(
                "running on a machine with at most 64 bit pointer width and can convert from \
                 usize to u64",
            ),
            tree_size: tree_size.try_into().expect(
                "running on a machine with at most 64 bit pointer width and can convert from \
                 usize to u64",
            ),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[cfg_attr(feature = "serde", serde(transparent))]
pub struct RollupId {
    #[cfg_attr(feature = "serde", serde(serialize_with = "crate::serde::base64"))]
    inner: [u8; 32],
}

impl RollupId {
    /// Creates a new rollup ID from a 32 byte array.
    ///
    /// Use this if you already have a 32 byte array. Prefer
    /// [`RollupId::from_unhashed_bytes`] if you have a clear text
    /// name what you want to use to identify your rollup.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let bytes = [42u8; 32];
    /// let rollup_id = RollupId::new(bytes);
    /// assert_eq!(bytes, rollup_id.get());
    /// ```
    #[must_use]
    pub const fn new(inner: [u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner,
        }
    }

    /// Returns the 32 bytes array representing the rollup ID.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let bytes = [42u8; 32];
    /// let rollup_id = RollupId::new(bytes);
    /// assert_eq!(bytes, rollup_id.get());
    /// ```
    #[must_use]
    pub const fn get(self) -> [u8; 32] {
        self.inner
    }

    /// Creates a new rollup ID by applying Sha256 to `bytes`.
    ///
    /// Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// use sha2::{
    ///     Digest,
    ///     Sha256,
    /// };
    /// let name = "MyRollup-1";
    /// let hashed = Sha256::digest(name);
    /// let rollup_id = RollupId::from_unhashed_bytes(name);
    /// assert_eq!(rollup_id, RollupId::new(hashed.into()));
    /// ```
    #[must_use]
    pub fn from_unhashed_bytes<T: AsRef<[u8]>>(bytes: T) -> Self {
        Self {
            inner: Sha256::digest(bytes).into(),
        }
    }

    /// Allocates a vector from the fixed size array holding the rollup ID.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let rollup_id = RollupId::new([42u8; 32]);
    /// assert_eq!(vec![42u8; 32], rollup_id.to_vec());
    /// ```
    #[must_use]
    pub fn to_vec(self) -> Vec<u8> {
        self.inner.to_vec()
    }

    /// Convert a byte slice to a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_slice(bytes: &[u8]) -> Result<Self, IncorrectRollupIdLength> {
        let inner =
            <[u8; ROLLUP_ID_LEN]>::try_from(bytes).map_err(|_| IncorrectRollupIdLength {
                received: bytes.len(),
            })?;
        Ok(Self::new(inner))
    }

    /// Converts a byte vector to a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_vec(bytes: Vec<u8>) -> Result<Self, IncorrectRollupIdLength> {
        let inner =
            <[u8; ROLLUP_ID_LEN]>::try_from(bytes).map_err(|bytes| IncorrectRollupIdLength {
                received: bytes.len(),
            })?;
        Ok(Self::new(inner))
    }

    #[must_use]
    pub fn to_raw(&self) -> raw::RollupId {
        raw::RollupId {
            inner: self.to_vec().into(),
        }
    }

    #[must_use]
    pub fn into_raw(self) -> raw::RollupId {
        raw::RollupId {
            inner: self.to_vec().into(),
        }
    }

    /// Converts from protobuf type to rust type for a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_raw(raw: &raw::RollupId) -> Result<Self, IncorrectRollupIdLength> {
        Self::try_from_slice(&raw.inner)
    }
}

impl AsRef<[u8]> for RollupId {
    fn as_ref(&self) -> &[u8] {
        &self.inner
    }
}

impl From<[u8; ROLLUP_ID_LEN]> for RollupId {
    fn from(inner: [u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner,
        }
    }
}

impl From<&[u8; ROLLUP_ID_LEN]> for RollupId {
    fn from(inner: &[u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner: *inner,
        }
    }
}

impl From<&RollupId> for RollupId {
    fn from(value: &RollupId) -> Self {
        *value
    }
}

impl std::fmt::Display for RollupId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Base64Display::new(self.as_ref(), &BASE64_STANDARD).fmt(f)
    }
}

#[derive(Debug, thiserror::Error)]
#[error("expected 32 bytes, got {received}")]
pub struct IncorrectRollupIdLength {
    received: usize,
}

/// Indicates that the protobuf response contained an array field that was not 20 bytes long.
#[derive(Debug, thiserror::Error)]
#[error("expected 20 bytes, got {received}")]
pub struct IncorrectAddressLength {
    received: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Address(
    #[cfg_attr(feature = "serde", serde(serialize_with = "crate::serde::base64"))]
    [u8; ADDRESS_LEN],
);

impl Address {
    #[must_use]
    pub fn get(self) -> [u8; ADDRESS_LEN] {
        self.0
    }

    #[must_use]
    pub fn to_vec(&self) -> Vec<u8> {
        self.0.to_vec()
    }

    /// Construct a sequencer address from a [`ed25519_consensus::VerificationKey`].
    ///
    /// The first 20 bytes of the sha256 hash of the verification key is the address.
    #[must_use]
    // Silence the clippy lint because the function body asserts that the panic
    // cannot happen.
    #[allow(clippy::missing_panics_doc)]
    pub fn from_verification_key(public_key: ed25519_consensus::VerificationKey) -> Self {
        /// this ensures that `ADDRESS_LEN` is never accidentally changed to a value
        /// that would violate this assumption.
        #[allow(clippy::assertions_on_constants)]
        const _: () = assert!(ADDRESS_LEN <= 32);
        let bytes: [u8; 32] = Sha256::digest(public_key).into();
        Self::try_from_slice(&bytes[..ADDRESS_LEN])
            .expect("can convert 32 byte hash to 20 byte array")
    }

    /// Convert a byte slice to an address.
    ///
    /// # Errors
    ///
    /// Returns an error if the account buffer was not 20 bytes long.
    pub fn try_from_slice(bytes: &[u8]) -> Result<Self, IncorrectAddressLength> {
        let inner = <[u8; ADDRESS_LEN]>::try_from(bytes).map_err(|_| IncorrectAddressLength {
            received: bytes.len(),
        })?;
        Ok(Self::from_array(inner))
    }

    #[must_use]
    pub const fn from_array(array: [u8; ADDRESS_LEN]) -> Self {
        Self(array)
    }

    #[must_use]
    pub fn to_raw(&self) -> raw::Address {
        raw::Address {
            inner: self.to_vec().into(),
        }
    }

    #[must_use]
    pub fn into_raw(self) -> raw::Address {
        raw::Address {
            inner: self.to_vec().into(),
        }
    }

    /// Convert from protobuf to rust type an address.
    ///
    /// # Errors
    ///
    /// Returns an error if the account buffer was not 20 bytes long.
    pub fn try_from_raw(raw: &raw::Address) -> Result<Self, IncorrectAddressLength> {
        Self::try_from_slice(&raw.inner)
    }
}

impl AsRef<[u8]> for Address {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

impl From<[u8; ADDRESS_LEN]> for Address {
    fn from(inner: [u8; ADDRESS_LEN]) -> Self {
        Self(inner)
    }
}

impl std::fmt::Display for Address {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Base64Display::new(self.as_ref(), &BASE64_STANDARD).fmt(f)
    }
}

/// Derive a [`merkle::Tree`] from an iterable.
///
/// It is the responsibility of the caller to ensure that the iterable is
/// deterministic. Prefer types like `Vec`, `BTreeMap` or `IndexMap` over
/// `HashMap`.
pub fn derive_merkle_tree_from_rollup_txs<'a, T: 'a, U: 'a>(rollup_ids_to_txs: T) -> merkle::Tree
where
    T: IntoIterator<Item = (&'a RollupId, &'a U)>,
    U: AsRef<[Vec<u8>]> + 'a + ?Sized,
{
    let mut tree = merkle::Tree::new();
    for (rollup_id, txs) in rollup_ids_to_txs {
        let root = merkle::Tree::from_leaves(txs.as_ref()).root();
        tree.build_leaf().write(rollup_id.as_ref()).write(&root);
    }
    tree
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;

    use super::{
        Address,
        IncorrectAddressLength,
    };

    #[test]
    fn account_of_20_bytes_is_converted_correctly() {
        let expected = Address([42; 20]);
        let account_vec = expected.0.to_vec();
        let actual = Address::try_from_slice(&account_vec).unwrap();
        assert_eq!(expected, actual);
    }

    #[track_caller]
    fn account_conversion_check(bad_account: &[u8]) {
        let error = Address::try_from_slice(bad_account);
        assert!(
            matches!(error, Err(IncorrectAddressLength { .. })),
            "converting form incorrect sized account succeeded where it should have failed"
        );
    }

    #[test]
    fn account_of_incorrect_length_gives_error() {
        account_conversion_check(&[42; 0]);
        account_conversion_check(&[42; 19]);
        account_conversion_check(&[42; 21]);
        account_conversion_check(&[42; 100]);
    }

    #[test]
    fn snapshots() {
        let address = Address([42; 20]);
        assert_json_snapshot!(address);
    }
}
