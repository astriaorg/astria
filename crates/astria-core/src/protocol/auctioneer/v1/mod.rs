use astria_core_address::Address;

use crate::{
    generated::astria::protocol::auctioneer::v1::EnshrinedAuctioneerEntry as RawEnshrinedAuctioneerEntry,
    primitive::v1::{
        asset,
        AddressError,
    },
    Protobuf,
};

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct EnshrinedAuctioneerEntryError(EnshrinedAuctioneerEntryErrorKind);

#[derive(Debug, thiserror::Error)]
enum EnshrinedAuctioneerEntryErrorKind {
    #[error("the expected field in the raw source type was not set: `{0}`")]
    FieldNotSet(&'static str),
    #[error("the `to` field was invalid")]
    Address { source: AddressError },
    #[error("the `fee_asset` field was invalid")]
    FeeAsset { source: asset::ParseDenomError },
    #[error("the `asset` field was invalid")]
    Asset { source: asset::ParseDenomError },
}

impl EnshrinedAuctioneerEntryError {
    #[must_use]
    fn field_not_set(field: &'static str) -> Self {
        Self(EnshrinedAuctioneerEntryErrorKind::FieldNotSet(field))
    }

    #[must_use]
    fn address(source: AddressError) -> Self {
        Self(EnshrinedAuctioneerEntryErrorKind::Address {
            source,
        })
    }

    #[must_use]
    fn fee_asset(source: asset::ParseDenomError) -> Self {
        Self(EnshrinedAuctioneerEntryErrorKind::FeeAsset {
            source,
        })
    }

    #[must_use]
    fn asset(source: asset::ParseDenomError) -> Self {
        Self(EnshrinedAuctioneerEntryErrorKind::Asset {
            source,
        })
    }
}

#[derive(Debug, thiserror::Error)]
#[must_use]
enum EnshrineAuctioneerErrorKind {
    #[error("the expected field in the raw source type was not set: `{0}`")]
    FieldNotSet(&'static str),
    #[error("the `to` field was invalid")]
    Address { source: AddressError },
    #[error("the `fee_asset` field was invalid")]
    FeeAsset { source: asset::ParseDenomError },
    #[error("the `asset` field was invalid")]
    Asset { source: asset::ParseDenomError },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EnshrinedAuctioneerEntry {
    pub auctioneer_address: Address,
    pub staker_address: Address,
    pub staked_amount: u128,
    pub fee_asset: asset::Denom,
    pub asset: asset::Denom,
}

impl Protobuf for EnshrinedAuctioneerEntry {
    type Error = EnshrinedAuctioneerEntryError;
    type Raw = RawEnshrinedAuctioneerEntry;

    #[must_use]
    fn into_raw(self) -> RawEnshrinedAuctioneerEntry {
        RawEnshrinedAuctioneerEntry {
            auctioneer_address: Some(self.auctioneer_address.into_raw()),
            staker_address: Some(self.staker_address.into_raw()),
            fee_asset: self.fee_asset.to_string(),
            asset: self.asset.to_string(),
            staked_amount: Some(self.staked_amount.into()),
        }
    }

    #[must_use]
    fn to_raw(&self) -> RawEnshrinedAuctioneerEntry {
        RawEnshrinedAuctioneerEntry {
            auctioneer_address: Some(self.auctioneer_address.to_raw()),
            staker_address: Some(self.staker_address.to_raw()),
            fee_asset: self.fee_asset.to_string(),
            asset: self.asset.to_string(),
            staked_amount: Some(self.staked_amount.into()),
        }
    }

    /// Convert from a raw, unchecked protobuf
    /// [`crate::generated::astria::protocol::transaction::v1::EnshrineAuctioneer`].
    ///
    /// # Errors
    ///
    /// - if the `auctioneer_address` field is not set
    /// - if the `staker_address` field is invalid
    /// - if the `fee_asset` field is invalid
    /// - if the `asset` field is invalid
    fn try_from_raw(proto: RawEnshrinedAuctioneerEntry) -> Result<Self, Self::Error> {
        let RawEnshrinedAuctioneerEntry {
            auctioneer_address,
            staker_address,
            fee_asset,
            asset,
            staked_amount,
        } = proto;
        let staked_amount =
            staked_amount.ok_or(EnshrinedAuctioneerEntryError::field_not_set("amount"))?;
        let auctioneer_address = auctioneer_address
            .ok_or_else(|| EnshrinedAuctioneerEntryError::field_not_set("auctioneer_address"))
            .and_then(|auctioneer_address| {
                Address::try_from_raw(auctioneer_address)
                    .map_err(EnshrinedAuctioneerEntryError::address)
            })?;
        let staker_address = staker_address
            .ok_or_else(|| EnshrinedAuctioneerEntryError::field_not_set("staker_address"))
            .and_then(|staker_address| {
                Address::try_from_raw(staker_address)
                    .map_err(EnshrinedAuctioneerEntryError::address)
            })?;
        let fee_asset = fee_asset
            .parse()
            .map_err(EnshrinedAuctioneerEntryError::fee_asset)?;
        let asset = asset
            .parse()
            .map_err(EnshrinedAuctioneerEntryError::asset)?;

        Ok(Self {
            auctioneer_address,
            staker_address,
            fee_asset,
            asset,
            staked_amount: staked_amount.into(),
        })
    }

    /// Convert from a reference to a raw, unchecked protobuf
    /// [`crate::generated::astria::protocol::transaction::v1::EnshrineAuctioneer`].
    ///
    /// # Errors
    ///
    /// - if the `auctioneer_address` field is not set
    /// - if the `staker_address` field is invalid
    /// - if the `fee_asset` field is invalid
    /// - if the `asset` field is invalid
    fn try_from_raw_ref(
        proto: &RawEnshrinedAuctioneerEntry,
    ) -> Result<Self, EnshrinedAuctioneerEntryError> {
        Self::try_from_raw(proto.clone())
    }
}
