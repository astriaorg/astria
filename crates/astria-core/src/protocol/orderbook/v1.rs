//! Order book protocol types for the Astria network.
//!
//! This module re-exports the auto-generated protobuf types from the Astria protocol
//! for the orderbook functionality.

// Re-export generated types
pub use crate::generated::astria::protocol::orderbook::v1::{
    Order, OrderMatch, Orderbook, OrderbookEntry, OrderSide, OrderTimeInForce, OrderType,
};
