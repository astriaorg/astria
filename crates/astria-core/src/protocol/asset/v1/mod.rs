use super::raw;
use crate::primitive::v1::asset::{
    self,
    Denom,
    ParseDenomError,
};

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct DenomResponseError(DenomResponseErrorKind);
impl DenomResponseError {
    #[must_use]
    fn invalid_denom(source: ParseDenomError) -> Self {
        Self(DenomResponseErrorKind::InvalidDenom {
            source,
        })
    }
}

#[derive(Debug, thiserror::Error)]
enum DenomResponseErrorKind {
    #[error("`denom` field was invalid")]
    InvalidDenom { source: ParseDenomError },
}

/// The sequencer response to a denomination request for a given asset ID.
#[derive(Clone, Debug, PartialEq)]
pub struct DenomResponse {
    pub height: u64,
    pub denom: Denom,
}

impl DenomResponse {
    /// Converts a protobuf [`raw::DenomResponse`] to an astria
    /// native [`DenomResponse`].
    ///
    /// # Errors
    /// Returns an error if the `denom` field  of the proto file can't be parsed as a [`Denom`].
    pub fn try_from_raw(proto: &raw::DenomResponse) -> Result<Self, DenomResponseError> {
        let raw::DenomResponse {
            height,
            denom,
        } = proto;
        Ok(Self {
            height: *height,
            denom: denom.parse().map_err(DenomResponseError::invalid_denom)?,
        })
    }

    /// Converts an astria native [`DenomResponse`] to a
    /// protobuf [`raw::DenomResponse`].
    #[must_use]
    pub fn into_raw(self) -> raw::DenomResponse {
        raw::DenomResponse::from_native(self)
    }
}

impl raw::DenomResponse {
    /// Converts an astria native [`DenomResponse`] to a
    /// protobuf [`raw::DenomResponse`].
    #[must_use]
    pub fn from_native(native: DenomResponse) -> Self {
        let DenomResponse {
            height,
            denom,
        } = native;
        Self {
            height,
            denom: denom.to_string(),
        }
    }

    // /// Converts a protobuf [`raw::DenomResponse`] to an astria
    // /// native [`DenomResponse`].
    // #[must_use]
    // pub fn try_into_native(self) -> Result<DenomResponse, DenomResponseError> {
    //     DenomResponse::try_from_raw(&self)
    // }

    // /// Converts a protobuf [`raw::DenomResponse`] to an astria
    // /// native [`DenomResponse`] by allocating a new [`v1::DenomResponse`].
    // #[must_use]
    // pub fn to_native(&self) -> Result<DenomResponse, DenomResponseError> {
    //     DenomResponse::try_from_raw(self)
    // }
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct AllowedFeeAssetsResponseError(AllowedFeeAssetsResponseErrorKind);

#[derive(Debug, thiserror::Error)]
enum AllowedFeeAssetsResponseErrorKind {
    #[error("failed to parse assets as IBC ICS20 denom")]
    IncorrectAsset(#[source] ParseDenomError),
}

impl AllowedFeeAssetsResponseError {
    fn incorrect_asset(inner: ParseDenomError) -> Self {
        Self(AllowedFeeAssetsResponseErrorKind::IncorrectAsset(inner))
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct AllowedFeeAssetsResponse {
    pub height: u64,
    pub fee_assets: Vec<asset::Denom>,
}

impl AllowedFeeAssetsResponse {
    /// Converts a protobuf [`raw::AllowedFeeAssetsResponse`] to an astria
    /// native [`AllowedFeeAssetsResponse`].
    ///
    /// # Errors
    /// - If one of the fee asset strings cannot be parsed as an [`asset::Denom`].
    pub fn try_from_raw(
        proto: &raw::AllowedFeeAssetsResponse,
    ) -> Result<Self, AllowedFeeAssetsResponseError> {
        let raw::AllowedFeeAssetsResponse {
            height,
            fee_assets,
        } = proto;
        let mut assets: Vec<asset::Denom> = Vec::new();

        for s in fee_assets {
            let native = s
                .parse()
                .map_err(AllowedFeeAssetsResponseError::incorrect_asset)?;
            assets.push(native);
        }

        Ok(Self {
            height: *height,
            fee_assets: assets,
        })
    }

    /// Converts an astria native [`AllowedFeeAssetsResponse`] to a
    /// protobuf [`raw::AllowedFeeAssetsResponse`].
    #[must_use]
    pub fn into_raw(self) -> raw::AllowedFeeAssetsResponse {
        raw::AllowedFeeAssetsResponse::from_native(self)
    }
}

impl raw::AllowedFeeAssetsResponse {
    /// Converts an astria native [`AllowedFeeAssetsResponse`] to a
    /// protobuf [`raw::AllowedFeeAssetsResponse`].
    #[must_use]
    pub fn from_native(native: AllowedFeeAssetsResponse) -> Self {
        let AllowedFeeAssetsResponse {
            height,
            fee_assets,
        } = native;
        let fee_assets = fee_assets
            .into_iter()
            .map(|denom| denom.to_string())
            .collect();
        Self {
            height,
            fee_assets,
        }
    }

    /// Converts a protobuf [`raw::AllowedFeeAssetsResponse`] to an astria
    /// native [`AllowedFeeAssetsResponse`].
    ///
    /// # Errors
    /// - If one of the assets  cannot be parsed as an [`asset::Denom`].
    pub fn try_into_native(
        self,
    ) -> Result<AllowedFeeAssetsResponse, AllowedFeeAssetsResponseError> {
        AllowedFeeAssetsResponse::try_from_raw(&self)
    }

    /// Converts a protobuf [`raw::AllowedFeeAssetsResponse`] to an astria
    /// native [`AllowedFeeAssetsResponse`] by allocating a new
    /// [`v1alpha1::AllowedFeeAssetsResponse`].
    ///
    /// # Errors
    /// - If one of the assets  cannot be parsed as an [`asset::Denom`].
    pub fn try_to_native(&self) -> Result<AllowedFeeAssetsResponse, AllowedFeeAssetsResponseError> {
        self.clone().try_into_native()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn denom_response_from_raw_is_correct() {
        let raw = raw::DenomResponse {
            height: 42,
            denom: "nria".to_owned(),
        };
        let expected = DenomResponse {
            height: 42,
            denom: "nria".parse().unwrap(),
        };
        let actual = DenomResponse::try_from_raw(&raw).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn denom_response_into_raw_is_correct() {
        let native = DenomResponse {
            height: 42,
            denom: "nria".parse().unwrap(),
        };
        let expected = raw::DenomResponse {
            height: 42,
            denom: "nria".to_owned(),
        };
        let actual = native.into_raw();
        assert_eq!(expected, actual);
    }

    #[test]
    fn denom_response_roundtrip_is_correct() {
        let native = DenomResponse {
            height: 42,
            denom: "nria".parse().unwrap(),
        };
        let expected = native.clone();
        let actual = DenomResponse::try_from_raw(&native.into_raw()).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn allowed_fee_assets_try_from_raw_is_correct() {
        let raw = raw::AllowedFeeAssetsResponse {
            height: 42,
            fee_assets: vec![
                "asset_0".to_string(),
                "asset_1".to_string(),
                "asset_2".to_string(),
            ],
        };
        let expected = AllowedFeeAssetsResponse {
            height: 42,
            fee_assets: vec![
                "asset_0".parse().unwrap(),
                "asset_1".parse().unwrap(),
                "asset_2".parse().unwrap(),
            ],
        };
        let actual = AllowedFeeAssetsResponse::try_from_raw(&raw).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn allowed_fee_assets_into_raw_is_correct() {
        let native = AllowedFeeAssetsResponse {
            height: 42,
            fee_assets: vec![
                "asset_0".parse().unwrap(),
                "asset_1".parse().unwrap(),
                "asset_2".parse().unwrap(),
            ],
        };
        let expected = raw::AllowedFeeAssetsResponse {
            height: 42,
            fee_assets: vec![
                "asset_0".to_string(),
                "asset_1".to_string(),
                "asset_2".to_string(),
            ],
        };
        let actual = native.into_raw();
        assert_eq!(expected, actual);
    }

    #[test]
    fn allowed_fee_assets_roundtrip_is_correct() {
        let native = AllowedFeeAssetsResponse {
            height: 42,
            fee_assets: vec![
                "asset_0".parse().unwrap(),
                "asset_1".parse().unwrap(),
                "asset_2".parse().unwrap(),
            ],
        };
        let expected = native.clone();
        let actual = native.into_raw().try_into_native().unwrap();
        assert_eq!(expected, actual);
    }
}
