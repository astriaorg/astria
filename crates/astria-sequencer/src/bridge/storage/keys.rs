use astria_core::primitive::v1::RollupId;
use base64::{
    display::Base64Display,
    engine::general_purpose::URL_SAFE,
};

use crate::{
    accounts::AddressBytes,
    storage::keys::AccountPrefixer,
};

pub(in crate::bridge) const INIT_BRIDGE_ACCOUNT_BASE_FEE: &str = "bridge/init_account_fee";
pub(in crate::bridge) const BRIDGE_LOCK_BYTE_COST_MULTIPLIER: &str =
    "bridge/lock_byte_cost_multiplier";
pub(in crate::bridge) const BRIDGE_SUDO_CHANGE_FEE: &str = "bridge/sudo_change_fee";

pub(in crate::bridge) const BRIDGE_ACCOUNT_PREFIX: &str = "bridge/account/";
const BRIDGE_ACCOUNT_SUDO_PREFIX: &str = "bridge/sudo/";
const BRIDGE_ACCOUNT_WITHDRAWER_PREFIX: &str = "bridge/withdrawer/";

pub(in crate::bridge) const DEPOSITS_EPHEMERAL: &str = "bridge/deposits";
const DEPOSIT_PREFIX: &str = "bridge/deposit/";

/// Example: `bridge/account/gGhH....zZ4=/rollup_id`.
///                         |base64 chars|
pub(in crate::bridge) fn rollup_id<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/rollup_id",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
    )
}

/// Example: `bridge/account/gGhH....zZ4=/asset_id`.
///                         |base64 chars|
pub(in crate::bridge) fn asset_id<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/asset_id",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
    )
}

/// Example: `bridge/sudo/gGhH....zZ4=`.
///                      |base64 chars|
pub(in crate::bridge) fn bridge_account_sudo_address<T: AddressBytes>(address: &T) -> String {
    AccountPrefixer::new(BRIDGE_ACCOUNT_SUDO_PREFIX, address).to_string()
}

/// Example: `bridge/withdrawer/gGhH....zZ4=`.
///                            |base64 chars|
pub(in crate::bridge) fn bridge_account_withdrawer_address<T: AddressBytes>(address: &T) -> String {
    AccountPrefixer::new(BRIDGE_ACCOUNT_WITHDRAWER_PREFIX, address).to_string()
}

/// Example: `bridge/account/gGhH....zZ4=/withdrawal_event/<event id>`.
///                         |base64 chars|                |UTF-8 chars|
pub(in crate::bridge) fn bridge_account_withdrawal_event<T: AddressBytes>(
    address: &T,
    withdrawal_event_id: &str,
) -> String {
    format!(
        "{}/withdrawal_event/{}",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address),
        withdrawal_event_id
    )
}

pub(in crate::bridge) fn deposit(block_hash: &[u8; 32], rollup_id: &RollupId) -> String {
    format!(
        "{DEPOSIT_PREFIX}{}/{rollup_id}",
        Base64Display::new(block_hash, &URL_SAFE),
    )
}

pub(in crate::bridge) fn last_transaction_id_for_bridge_account<T: AddressBytes>(
    address: &T,
) -> String {
    format!(
        "{BRIDGE_ACCOUNT_PREFIX}{}/last_tx",
        Base64Display::new(address.address_bytes(), &URL_SAFE)
    )
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::Address;

    use super::*;

    const COMPONENT_PREFIX: &str = "bridge/";

    fn address() -> Address {
        "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
            .parse()
            .unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(INIT_BRIDGE_ACCOUNT_BASE_FEE);
        insta::assert_snapshot!(BRIDGE_LOCK_BYTE_COST_MULTIPLIER);
        insta::assert_snapshot!(BRIDGE_SUDO_CHANGE_FEE);
        insta::assert_snapshot!(DEPOSITS_EPHEMERAL);
        insta::assert_snapshot!(rollup_id(&address()));
        insta::assert_snapshot!(asset_id(&address()));
        insta::assert_snapshot!(bridge_account_sudo_address(&address()));
        insta::assert_snapshot!(bridge_account_withdrawer_address(&address()));
        insta::assert_snapshot!(bridge_account_withdrawal_event(&address(), "the-event"));
        insta::assert_snapshot!(deposit(&[1; 32], &RollupId::new([2; 32])));
        insta::assert_snapshot!(last_transaction_id_for_bridge_account(&address()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(INIT_BRIDGE_ACCOUNT_BASE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_LOCK_BYTE_COST_MULTIPLIER.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_SUDO_CHANGE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(DEPOSITS_EPHEMERAL.starts_with(COMPONENT_PREFIX));
        assert!(rollup_id(&address()).starts_with(COMPONENT_PREFIX));
        assert!(asset_id(&address()).starts_with(COMPONENT_PREFIX));
        assert!(bridge_account_sudo_address(&address()).starts_with(COMPONENT_PREFIX));
        assert!(bridge_account_withdrawer_address(&address()).starts_with(COMPONENT_PREFIX));
        assert!(
            bridge_account_withdrawal_event(&address(), "the-event").starts_with(COMPONENT_PREFIX)
        );
        assert!(deposit(&[1; 32], &RollupId::new([2; 32])).starts_with(COMPONENT_PREFIX));
        assert!(last_transaction_id_for_bridge_account(&address()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn bridge_account_prefix_should_be_prefix_of_relevant_keys() {
        assert!(rollup_id(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX));
        assert!(asset_id(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX));
        assert!(
            bridge_account_withdrawal_event(&address(), "the-event")
                .starts_with(BRIDGE_ACCOUNT_PREFIX)
        );
        assert!(
            last_transaction_id_for_bridge_account(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX)
        );
    }
}
