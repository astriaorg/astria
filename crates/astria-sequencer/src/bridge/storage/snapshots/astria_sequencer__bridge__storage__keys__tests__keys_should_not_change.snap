---
source: crates/astria-sequencer/src/bridge/storage/keys.rs
assertion_line: 95
expression: INIT_BRIDGE_ACCOUNT_BASE_FEE_KEY
---
bridge/init_account_fee
